core-primitives/stf-executor/src/executor_tests.rs

```diff
@@ -0,0 +1,317 @@
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use crate::{
+	error::{Error, Result},
+	executor::*,
+	traits::{StateUpdateProposer, StfExecuteGenericUpdate, StfExecuteTimedGettersBatch},
+};
+use codec::{Decode, Encode};
+use ita_stf::{
+	test_genesis::{endowed_account, test_genesis_setup, ENDOWED_ACC_FUNDS},
+	AccountId, Balance, ShardIdentifier, State, TrustedCall, TrustedGetter, TrustedGetterSigned,
+};
+use itp_ocall_api::EnclaveAttestationOCallApi;
+use itp_stf_state_handler::handle_state::HandleState;
+use itp_test::{
+	builders::parentchain_header_builder::ParentchainHeaderBuilder,
+	mock::{handle_state_mock::HandleStateMock, onchain_mock::OnchainMock},
+};
+use itp_types::H256;
+use sgx_externalities::SgxExternalitiesTrait;
+use sgx_tstd::panic;
+use sp_core::Pair;
+use sp_runtime::app_crypto::sp_core::blake2_256;
+use std::{sync::Arc, time::Duration, vec, vec::Vec};
+
+// FIXME: Create unit tests for update_states, execute_shield_funds, execute_trusted_call, execute_trusted_call_on_stf #554
+
+pub fn propose_state_update_executes_all_calls_given_enough_time() {
+	// given
+	let (stf_executor, ocall_api, state_handler) = stf_executor();
+	let mrenclave = ocall_api.get_mrenclave_of_self().unwrap().m;
+	let (_, shard) = init_state_and_shard_with_state_handler(state_handler.as_ref());
+	let sender = endowed_account();
+	let signed_call = TrustedCall::balance_transfer(
+		sender.public().into(),
+		sender.public().into(),
+		42,
+	)
+	.sign(&sender.clone().into(), 0, &mrenclave, &shard);
+	let call_operation_hash: H256 =
+		blake2_256(&signed_call.clone().into_trusted_operation(true).encode()).into();
+	let signed_call_two = TrustedCall::balance_transfer(
+		sender.public().into(),
+		sender.public().into(),
+		100,
+	)
+	.sign(&sender.clone().into(), 1, &mrenclave, &shard);
+	let call_operation_hash_two: H256 =
+		blake2_256(&signed_call_two.clone().into_trusted_operation(true).encode()).into();
+
+	let old_state_hash = state_hash(&state_handler.load_initialized(&shard).unwrap());
+
+	// when
+	let batch_execution_result = stf_executor
+		.propose_state_update(
+			&vec![signed_call.clone(), signed_call_two.clone()],
+			&ParentchainHeaderBuilder::default().build(),
+			&shard,
+			Duration::from_secs(1000),
+			|state| state,
+		)
+		.unwrap();
+
+	// then
+	assert_eq!(old_state_hash, batch_execution_result.state_hash_before_execution);
+	assert_eq!(batch_execution_result.executed_operations.len(), 2);
+	assert_eq!(
+		batch_execution_result.get_executed_operation_hashes(),
+		vec![call_operation_hash, call_operation_hash_two]
+	);
+	// Ensure that state has been updated and not actually written.
+	assert_ne!(
+		state_handler.load_initialized(&shard).unwrap(),
+		batch_execution_result.state_after_execution
+	);
```

Comment:
I think not equal tests with hashes are not bullet-proof. If possible, I'd prefer an `old_state_hash == get_state().hash()` test.

Reply:
On what basis do you say that? The implementation of `assert_ne` is basically just a `==` call?

Reply:
No, it is a `!=` call, which I think there can be many reasons for hashes to be not equal to each other. But yes, it does not only apply to hashes, actually. 😄

Reply:
Looking at the implementation, it uses `==`.

Reply:
Aha. Sorry, I was not clear. I did not mean that the 'implementation' isn't bulletproof. I meant in the same way as above that we could compare it with `==` regarding the post-import state hash.

Reply:
Ah sorry, now I get it. Testing for equality in the test is more specific, yes. In this case however, we just want to make sure the state wasn't update yet, so IMO the `ne` is enough (and makes the test smaller)